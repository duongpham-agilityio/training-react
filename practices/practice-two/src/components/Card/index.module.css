.card {
  height: 232px;
  display: flex;
  align-items: start;
  border-radius: 30px;
  background-color: var(--secondary-color);
  text-decoration: none;
  color: var(--primary-color);
  font-family: var(--primary-font);
  padding: 40px;
}

.cardItem {
  display: flex;
  flex-direction: column;
  gap: 10px;
}

.cardItem:first-child {
  flex: 1;
}

.cardItem:last-child {
  display: none;
}

.heading {
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  line-height: 26.4px;
}

.description {
  display: -webkit-box;
  display: box;
  -webkit-line-clamp: 2;
  line-clamp: 2;
  -webkit-box-orient: vertical;
  overflow: hidden;
  line-height: 24px;
}

.action {
  flex: 1;
  display: flex;
  gap: 10px;
  align-items: flex-end;
  padding: 20px 0 0;
  margin-top: -9px;
}

.time {
  font-size: 16px;
  font-weight: 400;
  text-align: right;
  line-height: 26.4px;
}

.image {
  width: 95px;
  height: 116px;
  background-color: white;
  border-radius: 7px;
}

/* Small */
@media (min-width: 576px) {
  .cardItem:last-child {
    display: flex;
  }
}

/* Medium */
@media (min-width: 768px) {
}

/* Large */
@media (min-width: 992px) {
}

/* Extra-large */
@media (min-width: 1200px) {
  .card {
    width: 455px;
    gap: 30px;
  }

  .cardItem {
    gap: 15px;
  }

  .time {
    font-size: 22px;
  }
}

/* Extra-extra-large */
@media (min-width: 1400px) {
  .card {
    gap: 60px;
  }

  .cardItem {
    gap: 20px;
  }
}
