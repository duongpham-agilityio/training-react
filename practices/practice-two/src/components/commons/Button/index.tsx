import { ButtonHTMLAttributes, memo } from 'react';

// Styles
import buttonStyles from 'components/commons/Button/index.module.css';

type Variant = 'default' | 'primary' | 'secondary' | 'danger';
type Size = 'small' | 'medium' | 'large';
type Border = 'b-md' | 'b-lg';
type Width = 'w-sm' | 'w-lg';
export type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
  label?: string;
  leftIcon?: string;
  rightIcon?: string;
  variant?: Variant;
  border?: Border;
  size?: Size;
  width?: Width;
};

const Icon = ({
  icon,
  subTitle = 'icon action',
  ...rest
}: {
  icon?: string;
  subTitle?: string;
}) =>
  icon ? (
    <img
      src={icon}
      alt={subTitle}
      className={buttonStyles.icon}
      {...rest}
      width="22px"
      height="25px"
    />
  ) : null;

const ButtonIcon = memo(Icon);

const Button = (props: ButtonProps) => {
  const {
    label = '',
    leftIcon = '',
    rightIcon = '',
    className = '',
    variant = 'default',
    size = 'medium',
    border = 'b-md',
    width = 'w-sm',
    type = 'button',
    ...rest
  } = props;

  const classes = `${buttonStyles.btn} ${buttonStyles[variant]}
  ${buttonStyles[size]} ${buttonStyles[border]}
  ${buttonStyles[width]} ${className}`;

  return (
    <button className={classes} type={type} {...rest}>
      <ButtonIcon
        icon={leftIcon}
        subTitle="This is the left icon for the button"
      />

      {label}

      <ButtonIcon
        icon={rightIcon}
        subTitle="This is the right icon for the button"
      />
    </button>
  );
};

export default memo(Button);
