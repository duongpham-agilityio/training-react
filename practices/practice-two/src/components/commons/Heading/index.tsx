import { HTMLAttributes, createElement, memo } from 'react';

// Styles
import stylesHeading from 'components/commons/Heading/index.module.css';

type Size = 'sm' | 'md' | 'lg' | 'xl';
type TagName = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
export type HeadingProps = HTMLAttributes<HTMLHeadingElement> & {
  label?: string;
  tag?: TagName;
  size?: Size;
};

const Heading = (props: HeadingProps) => {
  const {
    children,
    className,
    label = '',
    size = 'sm',
    tag = 'h2',
    ...rest
  } = props;
  const Component = createElement(
    tag,
    {
      ...rest,
      className: `${stylesHeading.heading} ${stylesHeading[size]} ${
        className ?? ''
      }`,
    },
    children || label
  );

  return Component;
};

export default memo(Heading);
