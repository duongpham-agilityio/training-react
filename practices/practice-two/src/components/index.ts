export {
  default as Notification,
  type NotificationProps,
} from './Notification';

export { default as Popup, type PopupProps } from './Popup';
export { default as Card, type CardProps } from './Card';
export { default as CardSkeleton } from './Card/Skeleton';
export { default as Error } from './Error';
