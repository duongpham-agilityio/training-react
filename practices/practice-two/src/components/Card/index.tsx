import { MouseEvent, memo } from 'react';
import { Link, LinkProps } from 'react-router-dom';

// Component
import { Button, Heading } from 'components/commons';

// Styles
import styles from 'components/Card/index.module.css';

// Assets
import NextIcon from 'assets/icons/next.svg';
import TrashIcon from 'assets/icons/trash.svg';

export type CardProps = Omit<LinkProps, 'to'> & {
  href?: string;
  title: string;
  description: string;
  publishedDate: string;
  imageUrl: string;
  onDeleteBook?: (_event: MouseEvent) => void;
};

const Card = (props: CardProps) => {
  const {
    title,
    description,
    publishedDate,
    imageUrl,
    href = '',
    onDeleteBook,
  } = props;

  return (
    <Link to={href} className={styles.card}>
      <div className={styles.cardItem}>
        <Heading label={title} className={styles.heading} />
        <p className={styles.description}>{description}</p>
        <div className={styles.action}>
          <Button leftIcon={NextIcon} variant="primary" />
          <Button
            leftIcon={TrashIcon}
            variant="danger"
            onClick={onDeleteBook}
          />
        </div>
      </div>
      <div className={styles.cardItem}>
        <p className={styles.time}>{publishedDate}</p>
        {imageUrl ? (
          <img src={imageUrl} alt={title} className={styles.image} />
        ) : (
          <div className={`${styles.image} ${styles.fake}`}></div>
        )}
      </div>
    </Link>
  );
};

export default memo(Card);
