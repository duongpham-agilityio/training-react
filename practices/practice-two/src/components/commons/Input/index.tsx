import {
  InputHTMLAttributes,
  MouseEvent,
  Ref,
  TextareaHTMLAttributes,
  forwardRef,
  memo,
} from 'react';

// Components
import Button from '../Button';

// Style
import inputStyles from 'components/commons/Input/index.module.css';
import textAreaStyle from 'components/commons/Input/textArea.module.css';

export type InputProps = InputHTMLAttributes<HTMLInputElement> &
  TextareaHTMLAttributes<HTMLTextAreaElement> & {
    variant?: 'input' | 'area';
    leftIcon?: string;
    rightIcon?: string;
    customRef?: Ref<HTMLInputElement | HTMLTextAreaElement>;
    leftIconClick?: (_event: MouseEvent) => void;
    rightIconClick?: (_event: MouseEvent) => void;
  };

const TextArea = (props: Omit<InputProps, 'variant'>) => {
  const { value = '', className = '', ...rest } = props;

  return (
    <textarea
      className={`${textAreaStyle.text} ${className}`}
      value={value}
      {...rest}
    ></textarea>
  );
};

const Input = (props: Omit<InputProps, 'variant'>) => {
  const {
    leftIcon = '',
    rightIcon = '',
    customRef,
    className = '',
    type = 'text',
    placeholder = 'Enter text...',
    leftIconClick,
    rightIconClick,
    ...rest
  } = props;

  return (
    <div className={`${inputStyles.input} ${className}`}>
      <Button
        leftIcon={leftIcon}
        style={{
          backgroundColor: 'transparent',
          padding: '0',
          width: 'fit-content',
          height: 'fit-content',
        }}
        onClick={leftIconClick}
        aria-label="Search button"
      />

      <input
        className={inputStyles.value}
        type={type}
        placeholder={placeholder}
        ref={customRef as Ref<HTMLInputElement>}
        {...rest}
      />

      <Button
        leftIcon={rightIcon}
        style={{
          backgroundColor: 'transparent',
          padding: '0',
          width: 'fit-content',
          height: 'fit-content',
        }}
        onClick={rightIconClick}
        aria-label="The right button"
      />
    </div>
  );
};

const Component = (
  { variant, ...rest }: Omit<InputProps, 'customRef'>,
  ref: Ref<HTMLInputElement | HTMLTextAreaElement>
) => {
  const component = {
    input: Input,
    area: TextArea,
  };

  const Render = component[variant ?? 'input'];

  return <Render {...rest} customRef={ref} />;
};

export default memo(forwardRef(Component));
