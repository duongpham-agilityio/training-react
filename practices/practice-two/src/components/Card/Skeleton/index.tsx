// Components
import { Button, Heading } from 'components/commons';

// Styles
import styles from 'components/Card/Skeleton/index.module.css';
import skeletonStyles from 'styles/commons/index.module.css';

const CardSkeleton = () => {
  return (
    <div
      className={`${styles.card} ${skeletonStyles.loading}`}
      data-testid="card-skeleton"
    >
      <div className={styles.cardItem}>
        <Heading
          className={`${styles.heading} ${skeletonStyles.loadingItem}`}
        />
        <p
          className={`${styles.description} ${skeletonStyles.loadingItem}`}
        ></p>
        <div className={styles.action}>
          <Button variant="primary" className={skeletonStyles.loadingItem} />
          <Button variant="danger" className={skeletonStyles.loadingItem} />
        </div>
      </div>
      <div className={styles.cardItem}>
        <p className={`${styles.time} ${skeletonStyles.loadingItem}`}></p>
        <div
          className={`${styles.image} ${styles.fake} ${skeletonStyles.loadingItem}`}
        ></div>
      </div>
    </div>
  );
};

export default CardSkeleton;
