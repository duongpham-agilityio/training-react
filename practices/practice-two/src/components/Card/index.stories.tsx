import { Meta, Story } from '@storybook/react';

// Components
import { Card, CardProps } from 'components';
import { BrowserRouter } from 'react-router-dom';

export const Template: Story<CardProps> = () => (
  <BrowserRouter>
    <Card
      title="HTML/CSS Ebook HTML/CSS Ebook HTML/CSS Ebook HTML/CSS Ebook"
      description="Description of some
  book will displayed here Description of some
  book will displayed here Description of some
  book will displayed here"
      publishedDate="9:00 AM"
      imageUrl=""
    />
  </BrowserRouter>
);

export default {
  title: 'Practice/components/Card',
  component: Card,
  decorators: [
    (Story) => (
      <div
        className=""
        style={{
          width: '1440px',
          display: 'grid',
          gridTemplateColumns: '1fr 1fr 1fr',
          gap: '22px',
          margin: '50px auto',
        }}
      >
        <Story />
        <Story />
      </div>
    ),
  ],
} as Meta;
