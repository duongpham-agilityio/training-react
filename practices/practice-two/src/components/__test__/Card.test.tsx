// Mocks
import 'components/__test__/mocks/image.test';

import { act } from 'react-dom/test-utils';
import { fireEvent, render } from '@testing-library/react';

// Components
import Card, { CardProps } from 'components/Card';
import { ButtonProps } from 'components/commons';

// Types
import { LinkProps } from 'react-router-dom';

jest.mock('components/commons/Button', () => ({
  default: ({ onClick }: ButtonProps) => (
    <button onClick={onClick}>Icon</button>
  ),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  Link: ({ children, to, className }: LinkProps) => (
    <a href={to.toString()} className={className}>
      {children}
    </a>
  ),
}));

const setup = (props: CardProps) => render(<Card {...props} />);

describe('Card', () => {
  it('Match snapshot', () => {
    const { container } = setup({
      description:
        'Lorem ipsum dolor sit amet consectetur adipisicing elit. Praesentium, dicta.',
      publishedDate: '1955-08-15',
      imageUrl:
        'http://books.google.com/books/content?id=51YEAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api',
      title: 'Lorem ipsum dolor',
    });

    expect(container).toMatchSnapshot();
  });

  it('Render with props', () => {
    const mockOnDeleteBook = jest.fn();
    const { getByRole, getAllByRole } = setup({
      description:
        'Lorem ipsum dolor sit amet consectetur adipisicing elit. Praesentium, dicta.',
      publishedDate: '1955-08-15',
      imageUrl: '',
      title: 'Lorem ipsum dolor',
      href: 'books/1',
      onDeleteBook: mockOnDeleteBook,
    });

    const linkElement = getByRole('link');
    const buttons = getAllByRole('button');
    expect(linkElement.getAttribute('href')).toBe('books/1');

    act(() => {
      fireEvent.click(buttons[1]);
    });

    expect(mockOnDeleteBook).toBeCalled();
  });
});
