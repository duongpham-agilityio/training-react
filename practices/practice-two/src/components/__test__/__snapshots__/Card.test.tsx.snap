// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Match snapshot 1`] = `
<div>
  <a
    href=""
  >
    <div>
      <h2
        class="undefined undefined "
      >
        Lorem ipsum dolor
      </h2>
      <p>
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Praesentium, dicta.
      </p>
      <div>
        <button>
          Icon
        </button>
        <button>
          Icon
        </button>
      </div>
    </div>
    <div>
      <p>
        1955-08-15
      </p>
      <img
        alt="Lorem ipsum dolor"
        src="http://books.google.com/books/content?id=51YEAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
      />
    </div>
  </a>
</div>
`;
